### PR TITLE
Fix istio cert rotation bug for issue #4744

### PR DESCRIPTION
**Pull request by jonathanelbailey**
## Description

This PR fixes a known issue when Emissary is configured to connect to an Istio mTLS network.  When Emissary's Istio sidecar initiates a cert rotation the `istio-certs` secret generates a cache entry, but does not generate a delta since the `istio-certs` cache entry does not map to an actual resource on the Kubernetes cluster.  This can eventually resolve itself if a delta is generated that can initiate a change, but other times it results in an unrecoverable error that can cause traffic disruption without custom health checks.

## Related Issues

https://github.com/emissary-ingress/emissary/issues/4744

## Testing

* Replicated the issue using recommended steps from https://github.com/emissary-ingress/emissary/issues/4744#issuecomment-1406798652
* Using this [test image](https://hub.docker.com/r/jebaile7964/emissary) that includes the fix, the test case could no longer be replicated
* Tested on Istio 1.23.2
* Added parametrized unit tests for improved coverage of this test case

## Checklist

- [ ] **Does my change need to be backported to a previous release?**
  - What backport versions were discussed with the Maintainers in the Issue?

- [ ] **I made sure to update `CHANGELOG.md`.**

   Remember, the CHANGELOG needs to mention:
  - Any new features
  - Any changes to our included version of Envoy
  - Any non-backward-compatible changes
  - Any deprecations

- [x] **This is unlikely to impact how Ambassador performs at scale.**

   Remember, things that might have an impact at scale include:
  - Any significant changes in memory use that might require adjusting the memory limits
  - Any significant changes in CPU use that might require adjusting the CPU limits
  - Anything that might change how many replicas users should use
  - Changes that impact data-plane latency/scalability

- [ ] **My change is adequately tested.**

   Remember when considering testing:
  - Your change needs to be specifically covered by tests.
    - Tests need to cover all the states where your change is relevant: for example, if you add a behavior that can be enabled or disabled, you'll need tests that cover the enabled case and tests that cover the disabled case. It's not sufficient just to test with the behavior enabled.
  - You also need to make sure that the _entire area being changed_ has adequate test coverage.
    - If existing tests don't actually cover the entire area being changed, add tests.
    - This applies even for aspects of the area that you're not changing – check the test coverage, and improve it if needed!
  - We should lean on the bulk of code being covered by unit tests, but...
  - ... an end-to-end test should cover the integration points

- [ ] **I updated `CONTRIBUTING.md` with any special dev tricks I had to use to work on this code efficiently.**

- [ ] **The changes in this PR have been reviewed for security concerns and adherence to security best practices.**

### CODE REVIEW

**Comment by kflynn** [comment]
Hey @jonathanelbailey, are you on the CNCF Slack by any chance? I'd like to chat about this one... 🙂

File: python/tests/unit/test_ir.py [hunk 1]
```diff
@@ -58,3 +67,50 @@ class TestCase:
...
+        (
+            "null-cache-without-deltas",
+            None,
+            None,
+            {"config_type": "complete", "reset_cache": True, "invalidate_groups_for": []}
+        )
+        
+        ]
+    )
+    def test_check_deltas(self, name, cache_entry, deltas, expected, caplog):
```
Comment (kflynn) [inline python/tests/unit/test_ir.py:100]: Silly question: does this test fail without your change? 🙂

**Review by kflynn — COMMENTED** [review]
I'm leaning toward taking this but there are some corner cases I'm thinking about -- nice work in any case, many thanks! 🙂 I'm most curious, I think, about the test case, and about what testing you've put this through in general?

Thanks!! and happy to talk on Slack if you'd rather.

File: python/ambassador/ir/ir.py [hunk 2]
```diff
@@ -135,56 +135,54 @@ def check_deltas(
         to_invalidate: List[str] = []
         invalidate_groups_for: List[str] = []
 
-        # OK. If we don't have a cache, just skip all this crap.
-        if cache is not None:
-            # We have a cache. Start by assuming that we'll need to reset it,
+        # OK. If we don't have a cache and there are no deltas, just skip all this crap.
+        if (cache and fetcher.deltas) is not None:
```
Comment (kflynn) [inline python/ambassador/ir/ir.py:139]: (Restoring the comment that @the-wondersmith accidentally deleted 😂):

This logic feels a little weird – overall I think you're going for 'if we have a cache and there are deltas", and I think that'd be better expressed as

```
if (cache is not None) and fetcher.deltas:
```

My reasoning is that `fetcher.deltas` actually cannot ever be `None` – it's either an empty list or a list with some things in it.

**Comment by ppeble** [comment]
FYI, I have attempted to address the comments here since we've been so delayed in responding ourselves. I have a new PR open here that contains those updates: https://github.com/emissary-ingress/emissary/pull/5935

I will likely close this in favor of the other one unless there are objections.

**Comment by ppeble** [comment]
I am closing this in favor of https://github.com/emissary-ingress/emissary/pull/5935. The other PR will contain these changes plus the updates asked for in the comments.
